resources/assets/lib/play-detail-menu.coffee

```diff
@@ -0,0 +1,82 @@
+###
+#    Copyright 2015-2018 ppy Pty. Ltd.
+#
+#    This file is part of osu!web. osu!web is distributed with the hope of
+#    attracting more community contributions to the core ecosystem of osu!.
+#
+#    osu!web is free software: you can redistribute it and/or modify
+#    it under the terms of the Affero GNU General Public License version 3
+#    as published by the Free Software Foundation.
+#
+#    osu!web is distributed WITHOUT ANY WARRANTY; without even the implied
+#    warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+#    See the GNU Affero General Public License for more details.
+#
+#    You should have received a copy of the GNU Affero General Public License
+#    along with osu!web.  If not, see <http://www.gnu.org/licenses/>.
+###
+
+import { createElement as el, createRef, PureComponent } from 'react'
+import { a, button, div, i } from 'react-dom-factories'
+
+export class PlayDetailMenu extends PureComponent
+  constructor: (props) ->
+    super props
+
+    @uuid = osu.uuid()
+    @menu = createRef()
+
+    @state =
+      active: false
+
+
+  componentDidMount: =>
+    $(document).on "click.#{@uuid}", @hide
+    $(document).on "keyup.#{@uuid}", @hide
+
+
+  componentWillUnmount: =>
+    $(document).off ".#{@uuid}", @hide
+
+
+  hide: (e) =>
+    event = e.originalEvent
+    if event.keyCode == 27 || (event.button == 0 && !(@menu.current in event.composedPath()))
```

Comment:
maybe add check for button as well otherwise the button doesn't toggle. Or move `@menu` ref instead.